### PR TITLE
Py3k add support for running the same codebase on python2.6/2.7 and 3.3+

### PR DESCRIPTION
**Pull request by Ormod**
This PR makes it possible to run and build Diamond with Python3. 

All unit tests on Python3 except for 3 (2 rabbitmq and 1 kafkatest) pass.
On Python2.x all the unittests pass as before.

I've tried to break up the changes into small logical, hopefully-easy-to-review incremental changes.

Note that this breaks compatility with Pythons before 2.6 but that has mostly been the de facto state for a lot of the collectors even before this. (they depend on things like subprocess.check_output)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,7 +3,8 @@ PROJECT=diamond
 VERSION :=$(shell bash version.sh )
 RELEASE :=$(shell ls -1 dist/*.noarch.rpm 2>/dev/null | wc -l )
 HASH	:=$(shell git rev-parse HEAD )
-DISTRO=precise
+DISTRO:=$(shell python -c "import platform;print(platform.linux_distribution()[0])")
+PYTHON?=python3
```
Comment (mattrobenolt) [inline Makefile]: I assume you were using this for testing, but I'd rather us default any `PYTHON` to `python2`, or at least just `python` and let the environment gods use whichever is default.

File: bin/diamond-setup [hunk 2]
```diff
@@ -77,7 +78,7 @@ def getCollectors(path):
 
 
 def typeToString(key):
-    if isinstance(obj.config[key], basestring):
```
Comment (mattrobenolt) [inline bin/diamond-setup]: Any of these `basestring` -> `str` translations have lost meaning.

This is covered in `six`, see: https://bitbucket.org/gutworth/six/src/8a545f4e906f6f479a6eb8837f31d03731597687/six.py?at=default#six.py-49

File: setup.py [hunk 3]
```diff
@@ -17,12 +17,8 @@ def running_under_virtualenv():
     return False
 
 
-if os.environ.get('USE_SETUPTOOLS'):
```
Comment (mattrobenolt) [inline setup.py]: Do you have any context on why this was a thing? Because I don't, which means, I'm a bit afraid to just drop it.

Is distutils not a thing at all in python3?

File: src/collectors/amavis/amavis.py [hunk 4]
```diff
@@ -75,6 +75,8 @@ def collect(self):
                 cmdline = [self.config['amavisd_exe'], '-c', '1']
             agent = subprocess.Popen(cmdline, stdout=subprocess.PIPE)
             agent_out = agent.communicate()[0]
+            if isinstance(agent_out, bytes):
+                agent_out = agent_out.decode("utf8")
```
Comment (mattrobenolt) [inline src/collectors/amavis/amavis.py]: Why was this added?

Comment (mattrobenolt) [inline src/collectors/amavis/amavis.py]: Oh, TIL `bytes` is an alias for `str` in python2.

Comment (mattrobenolt) [inline src/collectors/amavis/amavis.py]: Is it safe to assume `utf8` conversion will always work? What about `ascii`?

File: src/collectors/cpuacct_cgroup/cpuacct_cgroup.py [hunk 5]
```diff
@@ -62,8 +62,8 @@ def collect(self):
                 stat_file.close()
 
         # create metrics from collected utimes and stimes for cgroups
-        for parent, cpuacct in results.iteritems():
```
Comment (mattrobenolt) [inline src/collectors/cpuacct_cgroup/cpuacct_cgroup.py]: Can we keep the `iteritems` for python2? This is something handled by six.

File: src/collectors/darner/darner.py [hunk 6]
```diff
@@ -137,7 +137,7 @@ def collect(self):
                             queues[queue][queue_stat])
 
             # figure out what we're configured to get, defaulting to everything
-            desired = self.config.get('publish', stats.keys())
+            desired = self.config.get('publish', list(stats.keys()))
```
Comment (mattrobenolt) [inline src/collectors/darner/darner.py]: We can normalize on using `iterkeys()` here.

File: src/collectors/disktemp/disktemp.py [hunk 7]
```diff
@@ -67,7 +67,7 @@ def match_device(self, device, path):
             # If the regex has a capture group for pretty printing, pick
             # the last matched capture group
             if self.devices.groups > 0:
-                key = '.'.join(filter(None, [g for g in m.groups()]))
+                key = '.'.join([_f for _f in [g for g in m.groups()] if _f])
```
Comment (mattrobenolt) [inline src/collectors/disktemp/disktemp.py]: `if _f is not None` is more correct to the original meaning.

File: src/collectors/ip/ip.py [hunk 8]
```diff
@@ -103,7 +104,7 @@ def collect(self):
             data = data.split()
 
             # Zip up the keys and values
-            for i in xrange(1, len(header)):
```
Comment (mattrobenolt) [inline src/collectors/ip/ip.py]: plz to use `six.moves.range` instead so we can preserve the same behavior on python2 and 3.

File: src/collectors/mongodb/test/testmongodb.py [hunk 9]
```diff
@@ -6,13 +6,18 @@
...
-from mock import call
+from test import MagicMock
+from test import patch
+from test import call
 
 from diamond.collector import Collector
 from mongodb import MongoDBCollector
 
+try:
+    long
```
Comment (mattrobenolt) [inline src/collectors/mongodb/test/testmongodb.py]: fwiw, in python2, there's literally no reason to ever use `long` afaik. `int()` will automatically coerce to a `long` if the value is long enough. So we can just stop using `long()` entirely.

File: src/collectors/netscalersnmp/netscalersnmp.py [hunk 10]
```diff
@@ -215,8 +215,8 @@ def collect_snmp(self, device, host, port, community):
                                        community)[serviceTypeOid].strip("\'"))
 
             # Filter excluded service types
-            if serviceType in map(lambda v: int(v),
```
Comment (mattrobenolt) [inline src/collectors/netscalersnmp/netscalersnmp.py]: Why replace `map`?

File: src/collectors/ossec/ossec.py [hunk 11]
```diff
@@ -53,7 +53,9 @@ def collect(self):
         try:
             p = subprocess.Popen(command, stdout=subprocess.PIPE)
             res = p.communicate()[0]
-        except Exception, e:
+            if isinstance(res, bytes):
+                res = res.decode("utf8")
+        except Exception as e:
             self.log.error('Unable to exec cmd: %s, because %s'
                            % (' '.join(command), str(e)))
```
Comment (mattrobenolt) [inline src/collectors/ossec/ossec.py]: Doesn't `str` here have different meaning between python2/3?

File: src/diamond/pycompat.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# coding=utf-8
```
Comment (mattrobenolt) [inline src/diamond/pycompat.py]: We should just start by vendorizing the entirety of six for now imo. There are too many things we're using.

File: src/collectors/darner/darner.py (hunk 6 above)
Comment (Ormod) [inline src/collectors/darner/darner.py]: Iterkeys does not exist in python3

Comment (mattrobenolt) [inline src/collectors/darner/darner.py]: Sorry, I mean, normalize on what six does. They provide a wrapper to do the same behavior in both.

File: src/collectors/ossec/ossec.py (hunk 11 above)
Comment (Ormod) [inline src/collectors/ossec/ossec.py]: Not for the purposes of printing the exception name AFAIK.

**Comment by josegonzalez** [comment]
Nice work!

**Comment by varunpatro** [comment]
Any update on this? We use diamond heavily in our company and are now moving to python 3, so if I can help with this development, I'll be happy to help!

**Comment by shortdudey123** [comment]
@Ormod can you rebase this PR and address some of the feedback?

**Comment by Ormod** [comment]
I'm afraid we stopped using Diamond a while back but if someone else wants to work on this and re-use my code please feel free to do so.
